queries/nix.scm

```diff
@@ -0,0 +1,146 @@
+;; Statements
+
+[
+    (binding)
+] @statement
+;; This matches assert statements as stand-alone statements. This
+;; is because it's common to have a series of assert statements in a file, which
+;; would otherwise all match as a single statement. See the playground/nix/asserts.nix
+;; file for an example.
+(assert_expression
+    "assert" @statement.start
+    condition: (_)
+    ";" @statement.end
+    body: (_)
+)
+(binding_set) @map.iteration @list.iteration @ifStatement.iteration
+(source_code) @statement.iteration @map.iteration @comment.iteration
+
+;;!! let
+;;!!   a = 1;
+;;!    <*****
+;;!!   b = 2;
+;;!    *****>
+;;!! in a + b
+(let_expression
+    "let"
+    (_) @statement.iteration @value.iteration @name.iteration
+)
+
+;; Conditionals
+
+;;!! key = if a then b else c;
+;;!        ^^^^^^^^^^^^^^^^^^
+(if_expression) @ifStatement @branch.iteration
+;;!! key = if a then b else c;
+;;!        ^^^^
+(if_expression
+    "if" @branch.start
+    condition: (_) @branch.end
+)
+;;!! key = if a then b else c;
+;;!             ^^^^^^
+(if_expression
+    "then" @branch.start
+    consequence: (_) @branch.end
+)
+;;!! key = if a then b else c;
+;;!                    ^^^^^^
+(if_expression
+    "else" @branch.start
+    alternative: (_) @branch.end
+)
+
+;; Lists and maps
+
+;;!! foo = [ a b c ];
+;;!        ^^^^^^^^^
+(list_expression) @list
+
+;;!! foo = { x = 1; y = 2; };
+;;!        ^^^^^^^^^^^^^^^^^^
+[
+    (attrset_expression)
+    (rec_attrset_expression)
+] @map @statement.iteration @value.iteration @name.iteration
+
+;;!! foo = { x = 1; y = 2; };
+;;!          ^
+(attrpath) @collectionKey
+
+;; Strings
+
+;;!! # foo
+;;!  ^^^^^
+(comment) @comment @textFragment
+
+(string_expression
+    _ @textFragment.start.endOf
+    (string_fragment)
+    _ @textFragment.end.startOf
+) @string
+
+(indented_string_expression
+    "''" @textFragment.start.endOf
+    (string_fragment)
+    "''" @textFragment.end.startOf
+) @string
+
+;; Hack to test these two are otherwise sane selectors
+;; [
+;;     (string_expression)
+;;     (indented_string_expression)
+;; ] @comment
```

Comment:
what's this?

Reply:
Basically me trying to test to make string scope work, which it doesn't currently (why there are no string tests atm). I think it might be a cursorless bug with the text fallback as I get an error about not finding a pair. The " " string expression works, but I still think even in that case it's not using the query, as if in the other multi-line string (indented string expression, which is like python """ but with double single ticks: '' '') type I include a regular " " string inside it will pick that " " string up only, which is wrong. The comment hack was to be sure if I don't use @string the two string expression queries work correctly as I'd expect, which they do (also why I think it may be a cursorless bug in string matching).

Reply:
yeah the spoken "string" scope doesn't actually use the string scope type; it uses a surrounding pair scope. Is "change string" not working?

Reply:
No `change string` doesn't work. Basically strings will look like this:

```nix
{
  a = "double quoted string";
  b = ''two single quote string'';
  c = ''
    multi-line
    quote string'';
  d = "Interpolated ${value} string";
  e = "Escaped \${string} ";
  f = ''
    "Nested"
    string
    '';
}
```

However the only thing cursorless will pick up with the `change` command is the `"`-based ones, so `a`, `d`, and `e` assignments, and `"Nested"` (technically wrong). Anything `''` wrapped fails with `Couldn't find containing pair` error. My typescript fu is weak and so is my debugging in vscode, so I didn't get far in debugging it so far but it does seem to be related to falling back to text parsing.

It doesn't seem like there's no `visualize string` command, so it's also not clear how to visualize what it's finding.

Reply:
oh right yeah that's a tough one. Our surrounding pair scope is likely viewing each `''` as an empty string. Unfortunately, there's not a lot you can do about that for now. See https://github.com/cursorless-dev/cursorless/issues/1812#issuecomment-1691493746

Reply:
Cool, thanks for the context. I'm willing to try to fix it, so happy to discuss. On the one hand it would be nice to get any nix support in mostly works but on the other hand proper string handling seems like it would be super useful to have right away, so seems worth trying even if it takes a bit of time.

Reply:
Actually probably shouldn't be too hard; I'd have a look at where that map is referenced, and you should have access to the editor at that point, so you can do `editor.document.languageId` to get the language id and use a custom map if it's nix

Reply:
Okay that's fixed and pushed. The only thing is naming/organization:

- I called returning `delimiterToText` as `getSimpleDelimiterMap` to kind of mirror `complexDelimiterMap`. 

- I'm not sure is if you still want a `delimiterMap.ts` standalone file, and then a `getSimpleDelimiterMap.ts` only for that function. Now `delimiterToText` isn't referenced anywhere else, so seemed maybe okay to keep them together. 

- I also thought about adding `getComplexDelimiterMap()`, but atm because `complexDelimiterMap` only ever uses the keys from the `delimiterToText` it won't change even if the language has different values, so seemed unnecessary. These are all things I could guess at your preferences, but may as well ask instead.

- I also noticed `leftToRightMap` in that file isn't actually used anywhere so can be deleted I think, but also not sure about doing that as part of a totally unrelated code change, to keep commits clean.

Reply:
This is now tracked in PR #2012